plugin/procfs/procfs.cc

```diff
@@ -0,0 +1,527 @@
+    rest = std::string(query.begin() + first_sep_after_star + 1, query.end());
+  }
+
+  if (rest.size() == 0) {
+    return limited_glob_files(path, pattern, max_results);
+  }
+
+  DIR *dir = opendir(path.c_str());
+  struct dirent *dir_entry;
+
+  if (!dir) return files_found;
+
+  while ((dir_entry = readdir(dir)) != 0) {
+    if (!fnmatch(pattern.c_str(), dir_entry->d_name,
+                 FNM_CASEFOLD | FNM_NOESCAPE | FNM_PERIOD)) {
+      if (dir_entry->d_type == DT_DIR || dir_entry->d_type == DT_LNK ||
+          dir_entry->d_type == DT_UNKNOWN) {
+        std::string new_pattern = std::string(path)
+                                      .append("/")
+                                      .append(dir_entry->d_name)
+                                      .append("/")
+                                      .append(rest);
+        std::vector<std::string> v = limited_glob(
+            new_pattern, max_results - files_found.size(), max_depth - 1);
+        files_found.insert(files_found.end(), v.begin(), v.end());
+      }
+    }
+  }
+
+  closedir(dir);
+  return files_found;
+}
+
+static void fill_files_list(std::vector<std::string> &files) {
+  std::istringstream list(files_spec);
+
+  files.clear();
+  while (list) {
+    std::string path;
+    std::getline(list, path, ';');
+    if (path.rfind("/proc", 0) != 0 && path.rfind("/sys", 0) != 0) continue;
+
+    if (path.find_first_of("*?[") == std::string::npos) {
+      files.push_back(path);
+      continue;
+    }
+
+    std::vector<std::string> v1 =
+        limited_glob(path, MAX_FILES_PER_PATTERN, MAX_PATTERN_DEPTH);
+    files.insert(files.end(), v1.begin(), v1.end());
+  }
+}
+
+static void fill_view_row(THD *thd, TABLE *table, const char *fname, char *buf,
+                          size_t sz) {
+  if (sz == 0) return;
+
+  table->field[0]->store(fname, strlen(fname), system_charset_info);
+  table->field[1]->store(buf, sz, system_charset_info);
+  schema_table_store_record(thd, table);
+
+  files_read++;
+  bytes_read += sz;
+}
+
+static int fill_view(THD *thd, TABLE_LIST *tables, Item *cond) {
+  if (!thd->security_context()
+           ->has_global_grant(PRIVILEGE_NAME, strlen(PRIVILEGE_NAME))
+           .first) {
+    my_error(ER_SPECIFIC_ACCESS_DENIED_ERROR, MYF(0), PRIVILEGE_NAME);
+    access_violations++;
+    return 1;
+  }
+
+  TABLE *table = tables->table;
+
+  static const std::string I_S_FILE("INFORMATION_SCHEMA.PROCFS.FILE");
+  std::string like_arg;
+
+  std::map<std::string, bool> in_args;
+
+  if (cond != 0) {
+    std::string eq_arg;
+    if (!get_equal_condition_argument(cond, &eq_arg, I_S_FILE) &&
+        !eq_arg.empty()) {
+      in_args[eq_arg] = true;
+    } else if (!get_like_condition_argument(cond, &like_arg, I_S_FILE) &&
+               !like_arg.empty()) {
+    } else {
+      get_in_condition_argument(cond, in_args, I_S_FILE);
+    }
+  }
+
+  std::vector<std::string> files;
+  fill_files_list(files);
```

Comment:
Here we are passing the caller allocated vector to be filled in.
We also have limited_glob() and limited_glob_files() functions which effectively fill the vector as well, but in that cases we return local vector instead of passing caller-allocated one as the parameter and let the compiler to figure out that local vector should be moved to caller.

Just a matter of style, but within one file maybe we should stick to one way of another?

Reply:
yup, another good call, also a bit of an efficiency improvement to pass reference rather than return by value

Reply:
actually, no, now I see that there is a reason that these are done this way and that is to ultimately limit the number of found results.  It would still be possible to achieve this limiting with some re-writing, but, IMHO I would rather not go too deep down that hole at this time since this is already written and tested.